packages/remirror__extension-file/src/file-extension.tsx

```diff
@@ -0,0 +1,375 @@
+        fileHandler: (props) => {
+          let pos: number | undefined;
+
+          if (props.type === 'drop') {
+            pos = props.pos;
+          }
+
+          for (const file of props.files) {
+            this.startLoadFile(file, pos);
+          }
+
+          return true;
+        },
+      },
+    ];
+  }
+
+  @command()
+  uploadFile(files: File[]): CommandFunction {
+    return () => {
+      for (const file of files) {
+        this.startLoadFile(file);
+      }
+
+      return true;
+    };
+  }
+
+  @command()
+  updateFile(pos: number, attrs: FileAttributes): CommandFunction {
+    return ({ tr, dispatch }) => {
+      dispatch?.(tr.setNodeMarkup(pos, undefined, attrs));
+      return true;
+    };
+  }
+
+  @command()
+  insertFile(attributes: FileAttributes, selection?: PrimitiveSelection): CommandFunction {
+    return ({ tr, dispatch }) => {
+      const { from, to } = getTextSelection(selection ?? tr.selection, tr.doc);
+      const node = this.type.create(attributes);
+
+      dispatch?.(tr.replaceRangeWith(from, to, node));
+
+      return true;
+    };
+  }
+
+  @command()
+  deleteFile(pos: number): CommandFunction {
+    return ({ tr, state, dispatch }) => {
+      const node = state.doc.nodeAt(pos);
+
+      if (node && node.type === this.type) {
+        tr.delete(pos, pos + 1).scrollIntoView();
+        this.options.onDeleteFile({ tr, pos });
+        dispatch?.(tr);
+        return true;
+      }
+
+      return false;
+    };
+  }
+
+  @command()
+  renameFile(pos: number, fileName: string): CommandFunction {
+    return ({ tr, state, dispatch }) => {
+      const node = state.doc.nodeAt(pos);
+
+      if (node && node.type === this.type) {
+        dispatch?.(tr.setNodeMarkup(pos, undefined, { ...node.attrs, fileName }));
+        return true;
+      }
+
+      return false;
+    };
+  }
+
+  private preLoadFile(id: FileId, file: File, pos: number | undefined): Promise<FileLoader> | void {
+    const tr = this.store.view.state.tr;
+    const insertPos = insertFilePoint(tr.doc, isNumber(pos) ? pos : tr.selection.from, this.type);
+
+    if (!isNumber(insertPos)) {
+      // failed to find a postition to insert the file node
+      return;
+    }
+
+    // insert the file node
+    const attrs: FileAttributes = { ...this.options.getAttrsBeforeUpload(file), id };
+    tr.insert(insertPos, this.type.createChecked(attrs));
+
+    // create a fileLoader, which will read the file
+    const fileLoader: FileLoader = this.options.loadFileHandler(file);
+
+    // insert the placeholder decoration
+    setPlaceholderAction(tr, { add: { id, pos: insertPos, playload: fileLoader } });
+
+    this.store.view.dispatch(tr);
+
+    // return a promise that resolves when the file is loaded, whether
+    // successfully or unsuccessfully
+    return new Promise<FileLoader>((resolve) => {
+      fileLoader.addEventListener('loadend', () => {
+        resolve(fileLoader);
+      });
+    });
```

Comment:
Not sure this belongs here, think we should just return the `fileLoader` here, and do this in `startLoadFile` instead?